tools/vulture_whitelist/cockpit.py

```diff
@@ -0,0 +1,13 @@
+from cockpit.internal_endpoints import User
```

Comment:
I generated it by hand, although there's also `vulture --make-whitelist`, although it produces some pretty terrible output.

According to the docs https://github.com/jendrikseipp/vulture/blob/main/README.md the preferred method for marking stuff as used is by using whitelist files.  At first I shared your opinion but after a day of working with it, I started to agree.

I have to admit that for everything in the `cockpit.py` file, we should try to eliminate this.  I have an idea for a tweak to the `systemd_ctypes` API that would help with that, for example.

For the stuff in other libraries, I really like this form of indicating use.

And particularly for the "our stuff that's used in external git repositories" for `testlib.py` I'm really quite fond of this syntax.  It basically apes as if we did use it ourselves.  It's very intuitive.

There were a few cases where the whitelist files didn't feel right — mostly for stuff in tests which wasn't a part of any public API (our API, standard library, or other).  For those couple of cases I did use `# noqa` since it felt like the better option.

But taking the case of `Browser.get_checked()` as an example: writing some `# noqa this is used in cockpit-machines yada yada` seems much worse than containing that information in a separate file.

As mentioned in the docs, linked above:

> However, we recommend using whitelists instead of noqa comments, since noqa comments add visual noise to the code and make it harder to read.

... and after a day of playing around with this stuff, I pretty much agree.

Reply:
I still heavily disagree with this. testlib.py is external API, so some unused code needs to be tolerated. But if vulture finds dead code in src/cockpit/, then these are either a lot of false positives (in which case the test isn't very good), or it's actually dead code, in which case that whitelist failure just papers over it, and will prevent us from ever cleaning it up. `# noqa` comments of course don't really change that, but they at least make the vulture overrides explicit and keep that information at one place.

If you feel strongly that this is the right thing to do, I won't block the PR because of this, but I find this an unmaintainable and brittle hack.. (Sorry for the strong words, but I really feel that)

Reply:
So looking at the whitelist files, we have a few different categories:

 - standard library stuff that, arguably, should go upstream.  They have whitelist files for the standard library and are open to improvements.
 - two oddities from Gtk that probably will never belong properly anywhere, but could also possibly be replaced by using accessor functions instead of attributes
 - a handful of D-Bus properties on various objects which could be avoided by better `systemd_ctypes` API
 - external users of testlib API, which... well, we need some way of documenting "we want to keep this because other people are using it"...

We can iterate a bit on the above ideas, but eventually some things will have to stay.

I guess my challenge to you would be to come up with a better way of dealing with this stuff...

Reply:
The stdlib and Gtk bits are the ones I don't understand at all -- these are external libraries, why is there even a complaint about these? Shouldn't vulture only look at code which is defined in the given source directory? Enumerating random parts of the standard library seems entirely pointless to me. If it doesn't understand the source directory properly, then perhaps it could be called with a bunch of `--ignore-names` for the standard library classes, together with a pointer to an upstream bug report, as this is really just broken. Hopefully these can go away over time with a fixed vulture.

> I guess my challenge to you would be to come up with a better way of dealing with this stuff...

As I already said, I'd prefer some kind of `# noqa external-api` annotation for the testlib and cockpit.User classes. vulture can ignore a decorator, if that can be applied to whole classes, that'd be fine for me. Alternatively, `--ignore-names cockpit.User` etc. would actually be fine for this as well, with an "external API comment".